onnxruntime InferenceSession fix to bug #5308

ecm200 opened the pull request:
Some distributions of onnxruntime require the specification of the providers argument on calling
InferenceSession. E.g. onnxruntime-gpu.
The package import call does not differnetiate which architecture specific version has been installed,
as all are imported with onnxruntime.
onnxruntime documentation says that
from v1.9.0 some distributions require the providers list to be
provided on calling an InferenceSession.
Therefore the try catch nested structure below attempts to create an inference session
with just the model path as pre v1.9.0.
If that fails, it will use the providers list
 as part of the InferenceSession call.
At the moment this is just CUDA and CPU, and probably
should be expanded.

Fix for bug #5308 

## What changes are proposed in this pull request?

In the `onnx.py` module, the class `_OnnxModelWrapper` init method has been adapted to cater for the changes in how an ONNX model is loaded into an InferenceSession. Previously to v1.9.0 of onnxruntime, only the path to the model file was needed to start an inference session, however post v1.9.0, some builds require the explicit passing of a providers list.

The original implementation of loading the onnx model was:

```python
self.rt = onnxruntime.InferenceSession(path)
```

The following code shows the updated version, which if the original call fails, it falls back to providing a list of executor providers:

```python

try:
  self.rt = onnxruntime.InferenceSession(path)
except:
  try:
      self.rt = onnxruntime.InferenceSession(path, providers=providers)
  except:
      raise
```

## How is this patch tested?

I have tested this on pre and post v1.9.0 versions of the onnxruntime and onnxruntime-gpu packages, and this works in all instances.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [x] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [x] `integrations/azure`: Azure and Azure ML integrations
- [x] `integrations/sagemaker`: SageMaker integrations
- [x] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

harupy committed:
Fix lint issues (#5305)

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>
Signed-off-by: Ed Morris <ecm200@gmail.com>

dbczumar committed:
Remove num examples (#5304)

Signed-off-by: dbczumar <corey.zumar@databricks.com>
Signed-off-by: Ed Morris <ecm200@gmail.com>

harupy committed:
Run `apt-get update` before installing `libopenblas-dev` (#5313)

* run sudo update

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* fix typo

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

* fix typo again

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>
Signed-off-by: Ed Morris <ecm200@gmail.com>

edmorris200 committed:
onnxruntime InferenceSession fix
Some distributions of onnxruntime require the specification of the providers argument on calling
InferenceSession. E.g. onnxruntime-gpu.
The package import call does not differnetiate which architecture specific version has been installed,
as all are imported with onnxruntime.
onnxruntime documentation says that
from v1.9.0 some distributions require the providers list to be
provided on calling an InferenceSession.
Therefore the try catch nested structure below attempts to create an inference session
with just the model path as pre v1.9.0.
If that fails, it will use the providers list
 as part of the InferenceSession call.
At the moment this is just CUDA and CPU, and probably
should be expanded.

Signed-off-by: Ed Morris <ecm200@gmail.com>

edmorris200 committed:
onnxruntime InferenceSession fix
Some distributions of onnxruntime require the specification of the providers argument on calling
InferenceSession. E.g. onnxruntime-gpu.
The package import call does not differnetiate which architecture specific version has been installed,
as all are imported with onnxruntime.
onnxruntime documentation says that
from v1.9.0 some distributions require the providers list to be
provided on calling an InferenceSession.
Therefore the try catch nested structure below attempts to create an inference session
with just the model path as pre v1.9.0.
If that fails, it will use the providers list
 as part of the InferenceSession call.
At the moment this is just CUDA and CPU, and probably
should be expanded.

BenWilson2 commented on mlflow/onnx.py:
```diff
@@ -182,10 +182,27 @@ def _load_model(model_file):
+        #       At the moment this is just CUDA and CPU, and probably should be expanded. It's been added to the class init call as an optional argument for further user customisation.
+        try:
+            self.rt = onnxruntime.InferenceSession(path)
+        except:
```
Could you catch the appropriate exception here to initiate the alternate processing logic?

ecm200 replied:
So, if I understand you correctly, catch the exception explicitly by type and then call the alternative loading, instead of a nested try implementation that I have done? Did I understand you right?

In the case where the `providers` variable is needed, the exception thrown is a `ValueError`.

So the refactored version would look as follows:

```python
try:
   self.rt = onnxruntime.InferenceSession(path)
except ValueError:
   self.rt = onnxruntime.InferenceSession(path, providers=providers)
except:
   raise
```

BenWilson2 replied:
exactly! Just so that we can make sure we're only trying the other logic in the case of the exact type of error that we're expecting (Sorry, I was hoping that the onnx package had a custom exception other than the blanket ValueError, but it's generally a good practice to catch only the Exceptions that we want to catch to do any further processing).

ecm200 replied:
@BenWilson2 @dbczumar 

Refactor committed with the suggested changes catching the ValueError and then running the post v1.9.0 call with providers.

Added functionality in the `__init__` method of the `_OnnxModelWrapper` class. 
This now reads the `MLModel` file from the same directory as the `model.onnx` has been downloaded to. **The assumption is that the MLModel file will always be present, as the model will be logged using the onnx flavour**.
It attempts to find the field "providers" in the onnx model flavour fields of the MLModel yaml.
If does not find it (for backwards compatability), it uses the default list which is hardwired as an attribute `ONNX_EXECUTION_PROVIDES` of the onnx model flavour.

BenWilson2 commented on mlflow/onnx.py:
```diff
@@ -182,10 +182,27 @@ def _load_model(model_file):
+        if providers is None:
+            providers = ['CUDAExecutionProvider', 'CPUExecutionProvider']
+
+        # NOTE: Some distributions of onnxruntime require the specification of the providers argument on calling. E.g. onnxruntime-gpu.
```
Please adjust line length to 100 chars max.

ecm200 replied:
Will gladly adjust in the morning and submit commits to update this.

ecm200 replied:
On new commit.
Text notes have been limited to 100 characters and rewritten to according to the altered implementation.

BenWilson2 commented on mlflow/onnx.py:
```diff
@@ -182,10 +182,27 @@ def _load_model(model_file):
+
+        # NOTE: Some distributions of onnxruntime require the specification of the providers argument on calling. E.g. onnxruntime-gpu.
+        #       The package import call does not differnetiate which architecture specific version has been installed, as all are imported with onnxruntime.
+        #       onnxruntime documentation says that from v1.9.0 some distributions require the providers list to be provided on calling an InferenceSession.
```
Please add link in the docstring to their docs that discusses these providers and how they are defined / used

ecm200 replied:
Will gladly add in the morning and submit commits to update this.

ecm200 replied:
Links have been added to the `onnxruntime` descriptions of execution providers and to the specific API description of the `InferenceSession` class.

BenWilson2 commented on mlflow/onnx.py:
```diff
@@ -182,10 +182,27 @@ def _load_model(model_file):
 
 class _OnnxModelWrapper:
-    def __init__(self, path):
+    def __init__(self, path, providers=None):
```
for this to be accessible to an end-user, the `_load_pyfunc()` function would need to be updated to support loading with a particular intended end-use environment provider configuration which would in turn change the signature of `mlflow.pyfunc.load_model()` to make this a modifiable argument. 
This is probably not what we want to do here. 
We might want to think about setting this property for an onnx flavor in a different way (during logging of the model and have this property stored in the serialized artifact?) 
WDYT @harupy @dbczumar ?

ecm200 replied:
@BenWilson2 completely understand the reservations about changing the `mlflow.pyfunc.load_model()` functionality given it's application to all model flavours.

I think your suggestion about modifying the saving metadata information on saving of the model is quite a nice suggestion as this means that only the `onnx.py` module code would need to be changed, is that right?

So under the `model_flavours` section of the `MLmodel` yaml:

```yaml
flavors:
  onnx:
    data: model.onnx
    onnx_version: 1.10.2
    providers: ['CUDAExecutionProvider', CPUExecutionProvider']
```

From my assessment, that would be facilitated through an additional argument added to the `add_flavour` method call in line 140 of `onnx.py`:

From:

`mlflow_model.add_flavor(FLAVOR_NAME, onnx_version=onnx.__version__, data=model_data_subpath)`

to:

`mlflow_model.add_flavor(FLAVOR_NAME, onnx_version=onnx.__version__, data=model_data_subpath, providers=providers)`

Could the variable `providers` be an additional optional argument for the `save_model` function, giving the user the ability to specify their own preference of providers (with a default list if not specified)?

```python
def save_model(
    onnx_model,
    path,
    conda_env=None,
    mlflow_model=None,
    signature: ModelSignature = None,
    input_example: ModelInputExample = None,
    pip_requirements=None,
    extra_pip_requirements=None,
    providers=None
):
   if not providers:
      providers = ['CUDAExecutionProvider', 'CPUExecutionProvider']

```

dbczumar replied:
@ecm200 @BenWilson2 This proposal sounds great to me!

ecm200 replied:
@dbczumar @BenWilson2 

Great. I’ll draft a change and commit and update the PR.

Only thing I need to figure is how to get the info into the ONNX private class that handles the inference session setup. 

My proposal would be that given this class gets the path to the onnx model, then the expectation is that this model will always be an MLModel type format? Therefore we can expect to have an MLModel yaml file located in the same directory? Is that a correct assumption?

If that’s a reasonable assumption then the `__init__` method of the aforementioned class could simply load the model meta data which would have the `providers` parameters set. For backwards comparability, where the MLModel file does not have a entry for `providers` then the fallback option could be to have a default list which covers a reasonable range of execution types.

ecm200 replied:
New commit af21a11 has this implemented.
See below commetns.

dbczumar commented on mlflow/onnx.py:
```diff
@@ -182,10 +182,27 @@ def _load_model(model_file):
+        #       The package import call does not differnetiate which architecture specific version has been installed, as all are imported with onnxruntime.
+        #       onnxruntime documentation says that from v1.9.0 some distributions require the providers list to be provided on calling an InferenceSession.
+        #       Therefore the try catch nested structure below attempts to create an inference session with just the model path as pre v1.9.0.
+        #       If that fails, it will use the providers list call. 
```
@ecm200 Would it be possible to always instantiate the InferenceSession with `providers`, or should `providers` only be used as a fallback?

ecm200 replied:
@dbczumar That’s a very good point and only occurred to me last night.

Certainly v1.8.0 does have the variable but it is not required. I am not sure how far back the variable is present in the `onnxruntime` releases before that.

Potentially I think it’s easier to define it this way, as it’s a clearly defined breaking change in the `onnxruntime` documentation.

I suppose we could just pin to a minimum version of `onnxruntime` to `mlflow`?

So I’m erring to the current implementation but happy to do what you admins think is more appropriate or online with `mlflow` style and policy.

dbczumar replied:
@ecm200 Great point! In the interest of supporting a broader set of onnxruntime versions, I'm also in favor of the current implementation. Thank you for the detailed explanation.

ecm200 replied:
@dbczumar  New commit implements refactored approach as discussed.

edmorris200 committed:
Refactor onnxruntime call and providers list
onnxruntime call now uses specific ValueError catch.
Initial attempt at loading model uses just path to onnx file.
If ValueError, the v1.9.0 call with providers is used.
The providers list argument is now obtained from the MLModel file.
The MLModel file meta data is loaded with assumption it is in the same
directory as the onnx model file.
Updating of save_model() function needed.
Signed-off-by: Ed Morris <ecm200@gmail.com>

edmorris200 committed:
Refactor onnxruntime call and providers list
onnxruntime call now uses specific ValueError catch.
Initial attempt at loading model uses just path to onnx file.
If ValueError, the v1.9.0 call with providers is used.
The providers list argument is now obtained from the MLModel file.
The MLModel file meta data is loaded with assumption it is in the same
directory as the onnx model file.
Updating of save_model() function needed.
Signed-off-by: Ed Morris <ecm200@gmail.com>

Signed-off-by: Ed Morris <ecm200@gmail.com>

edmorris200 committed:
onnxruntime providers argument added to log model
For the onnx model flavour, added ability to provide onnxruntime
execution providers list.
onnx_execution_providors argument passed through log_model function
as kwargs.
Added to save_model function by kwargs pass through.
Model flavour metadata saving updated to providers field.
If not specified, saves default to file.
Tested and works for MLModel files with and without new field.
Signed-off-by: Ed Morris <ecm200@gmail.com>

Signed-off-by: Ed Morris <ecm200@gmail.com>

edmorris200 committed:
onnxruntime providers argument added to log model
For the onnx model flavour, added ability to provide onnxruntime
execution providers list.
onnx_execution_providors argument passed through log_model function
as kwargs.
Added to save_model function by kwargs pass through.
Model flavour metadata saving updated to providers field.
If not specified, saves default to file.
Tested and works for MLModel files with and without new field.
Signed-off-by: Ed Morris <ecm200@gmail.com>

ecm200 commented:
@BenWilson2 @dbczumar 

After commit af21a11.

Updated the `log_model` and `save_model` functions of the `mlflow.onnx` model flavor.

The `onnx_execution_providers` argument was passed through the `onnx.pyfunc.log_model` function by using the kwargs structure.

The `save_model` function of `mlflow.onnx` was updated to handle the additional argument. If the `onnx_execution_providers` variable remains uninitiated (None), then the variable is initialised with the `ONNX_EXECUTION_PROVIDERS` module parameter with default GPU then CPU settings.

The additional `providers` variable is supplied to the `onnx.Model.add_flavor` and appears in the resulting saved `MLModel` yaml configuration file for the uploaded model. It saves as follows:

```yaml
artifact_path: model/tracing_detectron2_mask_rcnn_R_50_FPN_1X__COCO_trained__model_zoo
flavors:
  onnx:
    data: model.onnx
    onnx_version: 1.8.1
    providers:
    - CUDAExecutionProvider
    - CPUExecutionProvider
  python_function:
    data: model.onnx
    env: conda.yaml
    loader_module: mlflow.onnx
    python_version: 3.8.12
model_uuid: 3efce24bba61429aa9543a6f48bff00b
run_id: 06491f3c2c1d422d98c0d45ddc75ffb0
signature:
```

The MLModel file has been successfully read in as part of the `mlflow.onnx` model loading function, and verified as picking up the `providers` list from the MLModel yaml file.

BenWilson2 commented:
Hi @ecm200 could you sign your commit (It's the DCO check that's failing in the Github Actions CI checks)?
You can set up your local git to automatically apply signed commits. See: [the docs on signing commits](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#sign-your-work) for more information and how to do this. 
For the currently running CI check, you can manually affirm that you're the author by click on that failed task to sign the commit being tested.

BenWilson2 commented on mlflow/onnx.py:
```diff
@@ -111,9 +114,17 @@ def save_model(
     :param extra_pip_requirements: {{ extra_pip_requirements }}
+    :param onnx_execution_providers: List of strings defining onnxruntime execution providers.
+                                     Defaults to example:
+                                     ['CUDAExecutionProvider', 'CPUExecutionProvider']
```
For docs rendering
```suggestion
                                     ``['CUDAExecutionProvider', 'CPUExecutionProvider']``
```

BenWilson2 commented on mlflow/onnx.py:
```diff
@@ -111,9 +114,17 @@ def save_model(
+                                     Defaults to example:
+                                     ['CUDAExecutionProvider', 'CPUExecutionProvider']
+                                     This uses GPU preferentially over CPU.
+                                     See onnxruntime API for further descriptions.
```
```suggestion
                                     See onnxruntime API for further descriptions:
                                     https://onnxruntime.ai/docs/execution-providers/
```

BenWilson2 commented on mlflow/onnx.py:
```diff
@@ -362,6 +414,11 @@ def log_model(
+                                     Defaults to example:
+                                     ['CUDAExecutionProvider', 'CPUExecutionProvider']
+                                     This uses GPU preferentially over CPU.
+                                     See onnxruntime API for further descriptions.
```
```suggestion
                                     See onnxruntime API for further descriptions:
                                     https://onnxruntime.ai/docs/execution-providers/
```

BenWilson2 commented:
Once the linting is fixed and the commit is signed off LGTM!

ecm200 committed:
Code comment formatting changes

Code comments formatting corrections

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>

ecm200 committed:
Code comment formatting changes

Code comments formatting corrections

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Ed Morris <ecm200@gmail.com>

ecm200 committed:
Code comment formatting changes

Code formatting changes.

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Ed Morris <ecm200@gmail.com>

ecm200 committed:
Code comment formatting changes

Code formatting changes.

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>

ecm200 committed:
Code comment formatting changes

Code formatting changes

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>

ecm200 committed:
Code comment formatting changes

Code formatting changes

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Ed Morris <ecm200@gmail.com>

edmorris200 committed:
Black reformat changes to onnx.py
Ran BLACK to check formatting of ONNX.py
Signed-off-by: Ed Morris <ecm200@gmail.com>

Signed-off-by: Ed Morris <ecm200@gmail.com>

edmorris200 committed:
Black reformat changes to onnx.py
Ran BLACK to check formatting of ONNX.py
Signed-off-by: Ed Morris <ecm200@gmail.com>

edmorris200 committed:
Merge branch 'onnxruntime-InferenceSession-fix' of https://github.com/ecm200/mlflow into onnxruntime-InferenceSession-fix

ecm200 committed:
Merge branch 'onnxruntime-InferenceSession-fix' of https://github.com/ecm200/mlflow into onnxruntime-InferenceSession-fix

ecm200 commented:
@BenWilson2 @dbczumar 

I have tried retrospectively signing the unsigned commits, but it doesn't appear to have worked with all of them.

You mention an override button, but I believe I need write access to the repository in order for this to be shown.
I think this might be the easiest solution to fix it, but I can't access that button to do it unless I get write access.
Github docs say:

>Additionally, the DCO creates an override button accessible to only those with write access to the repository to create a successful check.

![image](https://user-images.githubusercontent.com/34489160/151780258-fd5af41b-df87-43c4-b9c2-2aa19f477ae6.png)

BenWilson2 commented on mlflow/onnx.py:
```diff
@@ -182,10 +199,49 @@ def _load_model(model_file):
+        except ValueError:
+            self.rt = onnxruntime.InferenceSession(path, providers=providers)
+        except:
+            raise
```
We don't need this final except catch block. If anything other than a ValueError raises, it'll throw an Exception normally. 
All that's needed here is:

```
try:
    self.rt = onnxruntime.InferenceSession(path)
except ValueError:
    self.rt = onnxruntime.InferenceSession(path, providers=providers)
```
Since we're catching the ValueError and handling it with a retry, we should be good to go with just that logic.

BenWilson2 commented on mlflow/onnx.py:
```diff
@@ -182,10 +199,49 @@ def _load_model(model_file):
         import onnxruntime
 
-        self.rt = onnxruntime.InferenceSession(path)
+        # Get the model meta data from the MLModel yaml file which may contain the providers specification
```
```suggestion
        # Get the model meta data from the MLModel yaml file which may contain the 
        # providers specification
```
To pacify the lint requirement of 100 lines

edmorris200 committed:
Removed catch all exception and linting issue.
Removed the catch all exception of the InferenceSession call.
Removed linting issue.
Signed-off-by: Ed Morris <ecm200@gmail.com>

BenWilson2 commented:
autoformat

mlflow-automation committed:
Autoformat: https://github.com/mlflow/mlflow/actions/runs/1779614231

Signed-off-by: mlflow-automation <mlflow-automation@users.noreply.github.com>

BenWilson2 reviewed (approved):
LGTM! 
Thank you very much for the contribution and I hope it's the first of many!

ecm200 commented:
> LGTM!
> Thank you very much for the contribution and I hope it's the first of many!

Happy to help and give back a little.

Do I need to do anything further to finish the PR?

I do have some thoughts on the ‘mlflow.azureml’ module and for potential to extend the functionality of the boilerplate entry script so that it supports more than just pandas dataframe tables, I’d like it to support the ‘TFServing’ inferencing format for Tensor type data so that it supports image data agnostically.

I’ve had some success editing the ‘mlflow.azureml.__init__.py’ file which contains the boilerplate entry script text as a string variable at the bottom of the file (line 657) . Replacing ‘parse_json_input’ with ‘ infer_and_parse_json_input’ would potentially go some of the way to solving that issue.

Should I scope some changes and put in another PR?

BenWilson2 commented:
For the ideas that you mentioned, feel free to create issues with feature requests (even if you're the one who will be working on them, it's a good medium for getting feedback / share design thoughts / canvas for other ideas in implementation). After the issue has received some comments then you can work on implementations :) 
The only thing left for this one is to merge it.

ecm200 commented:
> For the ideas that you mentioned, feel free to create issues with feature requests (even if you're the one who will be working on them, it's a good medium for getting feedback / share design thoughts / canvas for other ideas in implementation). After the issue has received some comments then you can work on implementations :) The only thing left for this one is to merge it.

@BenWilson2 this might be a dumb question, but I'll ask it just in case. 

I haven't actually done a PR into a repository I cannot edit before, so I am assuming the merge process is not on me, but either someone who does, or it is part of the CI process?

BenWilson2 commented:
I was in a meeting ;) 
It's merged now! Thanks again!
